lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,176 @@
+    if link, do: Process.link(pid)
+    {:reply, GenServer.call(state.local_name, {:subscribe, pid, topic}), state}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:unsubscribe, pid, topic}), state}
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:subscribers, topic}), state}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
+    redis_msg = {@redis_msg_vsn, state.node_ref, from_pid, topic, msg}
+    case :eredis.q(state.eredis_pid, ["PUBLISH", state.namespace, redis_msg]) do
+      {:ok, _}         -> {:reply, :ok, state}
+      {:error, reason} -> {:reply, {:error, reason}, state}
+    end
+  end
+
+  @doc """
+  Decodes binary message and fowards to local subscribers of topic
+  """
+  def handle_info({:message, _redis_topic, bin_msg, _cli_pid}, state) do
+    {_vsn, remote_node_ref, from_pid, topic, msg} = :erlang.binary_to_term(bin_msg)
+
+    if remote_node_ref == state.node_ref do
+      GenServer.call(state.local_name, {:broadcast, from_pid, topic, msg})
+    else
+      GenServer.call(state.local_name, {:broadcast, :none, topic, msg})
+    end
+
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _pid, {:connection_error, {:connection_error, :econnrefused}}}, state) do
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _linked_pid, _reason}, state) do
+    {:noreply, state}
+  end
+
+  @doc """
+  Connection establishment and shutdown loop
+
+  On init, an initial conection to redis is attempted when starting `:eredis_sub`.
+  If failed, the connection is tried again in `@reconnect_after_ms` until a max
+  of `@max_connect_attemps` is tried, at which point the server terminates with
+  an `:exceeded_max_conn_attempts`.
+  """
+  def handle_info(:establish_conn, %{reconnect_attemps: count} = state)
+    when count >= @max_connect_attemps do
+
+    {:stop, :exceeded_max_conn_attempts, state}
+  end
+  def handle_info(:establish_conn, state) do
+    handle_establish_conn(state)
+  end
+
+  def handle_info({:subscribed, _pattern, _client_pid}, state) do
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:eredis_connected, _client_pid}, state) do
+    Logger.info "redis connection re-established"
+    establish_success(state.eredis_sub_pid, state.eredis_pid, state)
+  end
+
+  def handle_info({:eredis_disconnected, _client_pid}, state) do
+    Logger.error "lost redis connection. Attempting to reconnect..."
+    {:noreply, %{state | status: :disconnected}}
+  end
+
+  def terminate(_reason, %{status: :disconnected}) do
+    :ok
+  end
+  def terminate(_reason, state) do
+    case :eredis_client.stop(state.eredis_sub_pid) do
+      :ok ->
+        case :eredis_client.stop(state.eredis_pid) do
```

Comment:
Maybe those cases should not be nested. Always shut both down regardless of the outcome? If you are worried about them being started, you may need to `catch :exit, {:noproc, _}`.